INT-4070: Fix @Gateway for Message receive style

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4070

The `MessagingGatewaySupport.receive()` is fully based on the `this.messagingTemplate.receiveAndConvert(replyChannel, null);`
which really just extracts `payload` from the `Message` via default `MessageConverter`.
Therefore `@Gateway` code, which expects to poll exactly `Message<?>` from the flow, is invalid at runtime with `ClassCastException`
- Expose `MessagingGatewaySupport.messagingTemplate` property as `protected` to give access for inheritors and siblings
- In the `GatewayProxyFactoryBean` use `messagingTemplate` and `replyChannel` directly from the `MethodInvocationGateway`
  to invoke raw `gateway.messagingTemplate.receive(replyChannel)` bypassing any conversion and to avoid breaking changes.
- Proof the solution with `GatewayProxyFactoryBean.testReceiveMessage()`

**Consider to cherry-pick (backport) down to 3.0.x**

artembilan committed:
INT-4070: Fix @Gateway for Message receive style

JIRA: https://jira.spring.io/browse/INT-4070

The `MessagingGatewaySupport.receive()` is fully based on the `this.messagingTemplate.receiveAndConvert(replyChannel, null);`
which really just extracts `payload` from the `Message` via default `MessageConverter`.
Therefore `@Gateway` code, which expects to poll exactly `Message<?>` from the flow, is invalid at runtime with `ClassCastException`

* Expose `MessagingGatewaySupport.messagingTemplate` property as `protected` to give access for inheritors and siblings
* In the `GatewayProxyFactoryBean` use `messagingTemplate` and `replyChannel` directly from the `MethodInvocationGateway`
to invoke raw `gateway.messagingTemplate.receive(replyChannel)` bypassing any conversion and to avoid breaking changes.
* Proof the solution with `GatewayProxyFactoryBean.testReceiveMessage()`

**Consider to cherry-pick (backport) down to 3.0.x**

artembilan commented:
See https://github.com/spring-projects/spring-integration-samples/pull/171 for real use-case.

garyrussell commented:
This is difficult to fix - and certainly not back-portable in its current state.

There is a possibility that someone is currently returning a `Message<?>` with a nested `Message<?>` payload; they would now get the outer message.

I suppose we could add a global SI property to drive the behavior and make the new behavior the default in 5.0. Any other ideas?

artembilan commented:
Gary, I don't understand your concern.
The code is like:
```
if (shouldReply) {
	if (shouldReturnMessage) {
		return gateway.receive();
	}
	response = gateway.receive();
}
```
even in `3.0`, but since we don't return `Message<?>` using `gateway.receive()` any more it is definitely bug in proposed API. And I'm not sure what you mean with "not back-portable in its current state", since the current code is like that everywhere, my fix should cherry-pick without conflicts. But, eh, I haven't tried to do that :smile: .

> someone is currently returning a Message<?> with a nested Message<?> payload

That sounds like exorcism and relying on bug, rather than just raise an issue against us, at least to worry that we provide strange behavior. If that, this guy will simply fix its code via:

```
Message<?> myMessage = gateway.receive();
myMessage = (Message<?>) myMessage.getPayload();
```

As you see I tried to provide as easy solution as possible without breaking changes for existing API.

We can add some global property, but I need more arguments why 
```
if (shouldReturnMessage) {
	return gateway.receive();
}
```
shound't return `Message<?>`.

garyrussell commented:
>relying on bug

Perhaps, but someone might reasonably surmise that it was designed that way - a gateway return type __always__ matches the `payload` type so, if I want the whole message, I need to wrap it. We simply can't change such behavior in a point release.

`shouldReturnMessage` simply means the method return type is `Message<?>` - we can't make any assumptions otherwise - the variable name is internal to the GW; the user knows nothing about it.

I know that was the intent, but the truth is that changing the behavior can break someone's app.

artembilan commented:
Eh... It really breaks my app. 

Having `@ServiceActivator` with the whole `Message<?>` as a method argument makes me feel that the same pattern is applied in case of `@Gateway` without any wrapping manipulation downstream before return. 

Another argument is request/reply. If we have `@Gateway` like:
```
Message<?> sendAndReceive(Message<?> message);
```
we end up with the:
```
if (shouldReply) {
	response = shouldReturnMessage ? gateway.sendAndReceiveMessage(args) : gateway.sendAndReceive(args);
}
```
which works as expected without any wrapping.

So, does why `Message<?> receive();` require from me wrapping when Framework should simply understand my wishes and return `Message<?>` as is?

garyrussell commented:
I am not saying you're wrong; you are completely correct; we just can't change the (default) behavior in `< 5.0`.

artembilan commented:
OK. Let me suggest you a gold middle.

We fix it only for `4.3.1` but with some global property (`spring.integration.messagingGateway.convertReceiveMessage=false` ?) and remove it in `5.0`.

The workaround for all other versions is like you said: wrap the target `Message<?>` into `payload` before sending to the gateway's `receiveChannel`.

Otherwise I'm almost convinced just to move it into `5.0` and wait for a demand to fix it in all other versions. :smile: 
My app can be simply fixed with the `String` return. But eh, we lost all those interesting Kafka headers...

garyrussell commented:
👍 works for me

artembilan commented:
OK. See https://jira.spring.io/browse/INT-4071
Will fix this PR accordingly today later.

artembilan committed:
Add `messagingGateway.convertReceiveMessage` global property to let revert to previous behavior

garyrussell commented on spring-integration-core/src/main/resources/META-INF/spring.integration.default.properties:
```diff
@@ -4,3 +4,4 @@ spring.integration.channels.maxBroadcastSubscribers=0x7fffffff
 spring.integration.taskScheduler.poolSize=10
 spring.integration.messagingTemplate.throwExceptionOnLateReply=false
 spring.integration.messagingAnnotations.require.componentAnnotation=false
+spring.integration.messagingGateway.convertReceiveMessage=false
```
True to retain current behavior?

artembilan replied:
Disagreed. 
I don't like to encourage end-users to rely on bug.
With the `false` we signal them that there is something wrong and has been fixed.
But if you don't like it, it's your responsibility to switch to `true`.

Otherwise they may think that we have many features which are just result of the bug :smile:

garyrussell replied:
There's a __big__ difference between "relying on a bug" and "working around a bug". This will break any such work-arounds. 

We should make life as painless as possible for someone who has had to already spend time and effort to find a work-around to a problem - (s)he may even have thought this is the way it's supposed to work and did not even consider it a workaround.

I therefore __strongly__ believe we should set it to `true` by default if we decide to have this flag.

Here's another solution that keeps the current behavior without the need for the property...

```java
if (shouldReturnMessage) {
    if (result.getPayload() instanceof Message) {
        return result.getPayload(); // unwrap for backwards compatibility
    else {
        return result;
    }
}

Perhaps log a warning (once) if we unwrap...

"In 5.0 the behavior for Message<?> return types will be such that a Message<?> wrapped in a Message<?> will not be unwrapped".

artembilan replied:
Your `if...else` looks wrong.
I tried to figure out the solution without the property, but it isn't going to go just with simple `if...else`.
We are expecting `Message<?>` but don't have control over `gateway.receive()`. Therefore always only `payload` independently of its `Message`. That's already target application responsibility to wrap message to another, e.g. to retain the headers for further logic.
I'm changing the property to `true` with the hope that no one will explore it. And that is much easier to me to fix my app, rather than answer to all those annoying questions like "Why have you fixed that bug?!"

artembilan committed:
* Make `messagingGateway.convertReceiveMessage=true` by default

* Mock `BeanFactory` in the `GatewayProxyFactoryBeanTests.testReceiveMessage()`
to be sure that `messagingGateway.convertReceiveMessage=false` works as expected

* Add `testReceiveMessageConvert()` to demonstrate `ClassCastException`

artembilan commented:
Gary, pay attention, please, to the new `testReceiveMessageConvert()`.
Maybe that `ClassCastException` just on call of `Message<String> getMessage();` will change your mind...

I'd say if anybody has caught such an exception before, he would notify us. No ? :cry:

garyrussell commented:
Yes, of course, I am just saying that it is possible that, instead of raising an issue with us, they thought it was working as designed and wrapped the message in another message as a solution.

It is perhaps ver unlikely that anybody has in fact done this, but I don't think we can take the chance.

artembilan committed:
Rebase and make `convertReceiveMessage` integration property as `false` by default

artembilan commented:
After some private discussion decided to make `spring.integration.messagingGateway.convertReceiveMessage=false`, since the root of cause is indeed an issue because it ends up with an exception:
```
assertThat(e, instanceOf(ClassCastException.class));
assertThat(e.getMessage(),
		containsString("java.lang.String cannot be cast to org.springframework.messaging.Message"));
```
which should be some kind of a bell for end-user anyway.
For all other exotic cases the `convertReceiveMessage` can be configured to `true` in the global `integration.properties`.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -452,7 +459,16 @@ private Object invokeGatewayMethod(MethodInvocation invocation, boolean runningO
+					else {
+						// TODO use messagingTemplate directly to avoid breaking changes
+						MessageChannel replyChannel = gateway.getReplyChannel();
+						Assert.state(replyChannel != null && (replyChannel instanceof PollableChannel),
```
unnecessary `null` check, `instanceof` is enough

artembilan replied:
Copy/paste artefact :smile:

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -452,7 +459,16 @@ private Object invokeGatewayMethod(MethodInvocation invocation, boolean runningO
+						return gateway.receive();
+					}
+					else {
+						// TODO use messagingTemplate directly to avoid breaking changes
```
Can you explain this?

artembilan replied:
See `MessagingGatewaySupport.receive()` implementation:
```
protected Object receive() {
     ...
	return this.messagingTemplate.receiveAndConvert(replyChannel, null);
}
```
So, we are not able to receive `Message` this way. Although canonical `GatewayProxyFactoryBean` logic expects to see it as a `Message`:
```
if (shouldReturnMessage) {
	return gateway.receive();
}
```
With that I assumed that original `MessagingGatewaySupport.receive()` implementation returned `Message` in ancient times. :smile: 
That's why I decided leave that method as is and use `messagingTemplate` directly.

Let me know if you'd prefer to see `MessagingGatewaySupport.receiveMessage()`, similarly with `sendAndReceiveMessage()`.

garyrussell replied:
No; I'm just wondering what the `TODO` is intended for.

artembilan committed:
Introduce `MessagingGatewaySupport.receiveMessage()` and use it from the `GatewayProxyFactoryBean`

garyrussell commented:
Merged as 5716315e8856323c9b6e0b02f7f4a2f4b8d615f3

robmoore commented:
I believe I may be running into this issue as I'm seeing the case with a message payload which is itself a message. I'm working with the 4.3.8 release, though, so wanted to confirm this fix had been backported as it sounds like 5.0 is where the fix was targeted.

artembilan commented:
@robmoore ,

Sorry, your concerns isn't clear or that is just my low English language level...

When we just to the [JIRA](https://jira.spring.io/browse/INT-4070) o the matter, we'll see the fix version as `4.3.1`. So, definitely this patch is present in the current `4.3.8` version.

Maybe when we have some test-case from your, we will be able to better understand what's going on.

Thanks

robmoore commented:
@artembilan 

Thanks for the quick reply. I may be off-base here but I'm seeing some puzzling behavior.

I have a flow defined like this:

```
    @Bean
    public IntegrationFlow searchIndex(ConnectionFactory connectionFactory, Advice retryAdvice) {
        return IntegrationFlows.from(Amqp.inboundAdapter(connectionFactory, "search-index").adviceChain(retryAdvice)
                .concurrentConsumers(4))
                .handle(message -> {
                    PatientSummary ps = (PatientSummary) message.getPayload();
                    try {
                        searchService.index(ps);
                    } catch (IOException e) {
                        logger.warn("Couldn't index patient summary: {}", ps);
                    }
                })
                .get();
    }

    @Bean
    public MessageChannel searchIndexChannel(ConnectionFactory connectionFactory) {
        return Amqp.channel(connectionFactory).queueName("search-index").get();
    }

    @MessagingGateway
    public interface SearchIndexGateway {
        @Gateway(requestChannel = "searchIndexChannel")
        void index(PatientSummary data);
    }
``` 
When I get to the line with the call to `message.getPayload()` I'm getting back a GenericMessage which contains the payload I expected to get with the call to `message.getPayload` initially. I'm unclear why the message-within-a-message scenario is occurring but it sounded like the issue you described above.

garyrussell commented:
What is sending to `searchIndexChannel`, and how?

artembilan commented:
No, that's not.

You see you use here `Amqp.channel(connectionFactory)`. And that is true the the whole Spring Integration Message is placed to the body of the AMQP message in this case:
```
/**
 * Set to true to extract the payload and map the headers; otherwise
 * the entire message is converted and sent. Default false.
 * @param extractPayload true to extract and map.
 * @since 4.3
 */
public void setExtractPayload(boolean extractPayload) {
```
since you don't listen from the `channel` via regular:
```
 IntegrationFlows.from("searchIndexChannel")
                .handle(message -> {
                    PatientSummary ps = (PatientSummary) message.getPayload();
```
but try to use `Amqp.inboundAdapter()` that isn't surprise that the whole `Message` comes as a `payload` from the AMQP message body mapping.

In this PR we talk about `@MessagingGateway` and `Message<?>` as return type of the gateway method.
So, your story isn't related.

Please, consider do not comment on the closed PRs. That is always fine to open JIRA/GH issue with bug or proposed improvement/feature. Everything rest must go to the community support resources like StackOverflow.

Thank you for understanding.

garyrussell commented:
This looks strange - you have two consumers on the `search-index` rabbit queue - the gateway and the inbound adapter. You should not use amqp-backed channels to distribute messages; they are intended for persistence only.
